### PR TITLE
Add setter for dataset_timestamp and model_performance_metrics

### LINKED ISSUE #979: performance metrics are not merged when merging DatasetProfileViews
## Description

DatasetProfileView's merge is missing some logic around fields outside of columns, need to refactor and fix this.

Notably, performance metrics are dropped while merging.

## Suggestions

Factor out functions for merging DatasetProfileView's metadata, metrics, columns, and timestamps.

## Related

Relates to organization/repo#number 

- [x] I have reviewed the [Guidelines for Contributing](https://github.com/whylabs/whylogs/blob/mainline/.github/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/whylabs/whylogs/blob/mainline/.github/CODE_OF_CONDUCT.md).

### PR DESCRIPTION
**Pull request by jamie256**
## Description
Updates for DatasetProfileView
* cleanup internal signature of performance metrics
* factor out merge logic for metrics, columns, metadata (new logic fixes merging perf metrics).
* add metadata
* add tests
Fixes #979

- [x] I have reviewed the [Guidelines for Contributing](CONTRIBUTING.md) and the [Code of Conduct](CODE_OF_CONDUCT.md).

### CODE REVIEW

**Review by andyndang — APPROVED** [review]
LGTM!

File: python/tests/core/view/test_dataset_profile_view.py [hunk 1]
```diff
@@ -138,10 +142,44 @@ def test_to_pandas_empty() -> None:
...
+
+
+def test_merge_takes_oldest_timestamps() -> None:
+    view1 = DatasetProfile().view()
+    view2 = DatasetProfile().view()
+    assert view1.dataset_timestamp < view2.dataset_timestamp
+    assert view1.creation_timestamp < view2.creation_timestamp
+    merged_view = view2.merge(view1)
+    assert merged_view.creation_timestamp == view1.creation_timestamp
+    assert merged_view.dataset_timestamp == view1.dataset_timestamp
```
Comment (richard-rogers) [inline python/tests/core/view/test_dataset_profile_view.py:163]: Did we decide we wanted this to be commutative, i.e., you'd get the same timestamps with `merged_view = view1.merge(view2)` ?

File: python/tests/core/view/test_dataset_profile_view.py [hunk 2]
```diff
@@ -138,10 +142,44 @@ def test_to_pandas_empty() -> None:
...
+
+
+def test_datasetprofile_zero_timestamps() -> None:
+    view = DatasetProfileView.zero()
+    stamp = time.time()
+    assert view is not None
+    assert view.creation_timestamp is None
+    assert view.dataset_timestamp is None
+    view.dataset_timestamp = stamp
+    assert view.dataset_timestamp is not None
```
Comment (richard-rogers) [inline python/tests/core/view/test_dataset_profile_view.py:173]: Why these last 2 asserts?

File: python/tests/core/view/test_dataset_profile_view.py [hunk 3]
```diff
@@ -138,10 +142,44 @@ def test_to_pandas_empty() -> None:
...
+    assert view.dataset_timestamp == stamp
+
+
+def test_datasetprofile_model_performance_metrics() -> None:
+    view = DatasetProfileView.zero()
+    assert view.model_performance_metrics is None
+    perf_metric = ModelPerformanceMetrics()
+    view.model_performance_metrics = perf_metric
+    assert view.model_performance_metrics is perf_metric
```
Comment (richard-rogers) [inline python/tests/core/view/test_dataset_profile_view.py:182]: Why this last assert?

File: python/tests/core/view/test_dataset_profile_view.py (hunk 2 above)
Comment (jamie256) [reply python/tests/core/view/test_dataset_profile_view.py:173]: exercising the property setter I added, and showing that DatasetProfileView.zero() has unset timestamps.

**Review by bernease — APPROVED** [review]
Stepped through this for my own understanding, but it LGTM!

File: python/tests/core/view/test_dataset_profile_view.py (hunk 1 above)
Comment (jamie256) [reply python/tests/core/view/test_dataset_profile_view.py:163]: They aren't exactly:
* They will be same error bounds but not guaranteed to be the same estimates on probabilistic algorithms (sketches)
* Where information is not mergeable (metadata key collisions), left hand side wins. So in practice you probably don't need to think too much about the order but technically it does make a difference.

File: python/tests/core/view/test_dataset_profile_view.py (hunk 3 above)
Comment (jamie256) [reply python/tests/core/view/test_dataset_profile_view.py:182]: perf metrics can be large (confusion matrix), showing that they are added but not copied, exercising the new property setter

**Pull request merged**